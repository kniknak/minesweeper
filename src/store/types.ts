/**
 * cell status
 */
export enum CellState {
  /**
   * not yet checked
   */
  Default = 1,
  /**
   * checked
   */
  Revealed,
  Flagged,
}

/**
 * game status
 */
export enum GameState {
  /**
   * initial/not yet started
   */
  Default = 1,
  Started,
  Failed,
  Succeeded,
}

/**
 * store state
 */
export type MinesweeperState = {
  /**
   * field width
   */
  width: number;
  /**
   * field height
   */
  height: number;
  /**
   * mines count on the field
   */
  minesCount: number;
  /**
   * current game state
   */
  gameState: GameState;
  /**
   * state state in the current game
   */
  cells: CellState[];
  /**
   * placed mines in the current game
   */
  mines: boolean[];
  /**
   * number of mines around for each cell
   */
  minesAround: number[];
};
