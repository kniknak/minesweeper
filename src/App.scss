* {
  box-sizing: border-box;
}

.app {
  margin: 24px;
}

.face {
  font-size: 48px;
  text-align: center;
  cursor: pointer;
}