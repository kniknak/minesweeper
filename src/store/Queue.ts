export class Queue {
  private set = new Set<number>();
  private arr: number[] = [];

  push(id: number) {
    this.set.add(id);
    this.arr.push(id);
  }

  pop(): number {
    const id: number = this.arr.pop()!;

    this.set.delete(id);

    return id;
  }

  has(id: number) {
    return this.set.has(id);
  }

  get size(): number {
    return this.arr.length;
  }
}
