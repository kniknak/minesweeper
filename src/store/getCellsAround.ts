export const getCellsAround = (id: number, width: number, height: number): number[] => {
  const cells: number[] = [];

  const notInTopRow = id >= width;
  const notInBottomRow = id < (height - 1) * width;
  const notInLeftColumn = id % width;
  const notInRightColumn = id % width < width - 1;

  if (notInTopRow) {
    cells.push(id - width);

    if (notInLeftColumn) {
      cells.push(id - width - 1);
    }

    if (notInRightColumn) {
      cells.push(id - width + 1);
    }
  }

  if (notInBottomRow) {
    cells.push(id + width);

    if (notInLeftColumn) {
      cells.push(id + width - 1);
    }

    if (notInRightColumn) {
      cells.push(id + width + 1);
    }
  }

  if (notInLeftColumn) {
    cells.push(id - 1);
  }

  if (notInRightColumn) {
    cells.push(id + 1);
  }

  return cells;
};
