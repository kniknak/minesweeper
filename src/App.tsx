import React from "react";
import "./App.scss";
import { Field } from "./Field/Field";
import { useDispatch, useSelector } from "react-redux";
import { AppDispatch, minesweeper } from "./store/store";
import { Settings } from "./Settings/Settings";
import { GameState, MinesweeperState } from "./store/types";

const gameStateToSmile = {
  [GameState.Default]: "ðŸ™‚",
  [GameState.Started]: "ðŸ™‚",
  [GameState.Failed]: "ðŸ˜µ",
  [GameState.Succeeded]: "ðŸ˜Ž",
};

export const App: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const gameState = useSelector<MinesweeperState, GameState>((state) => state.gameState);

  return (
    <div className="app">
      <Settings />
      <div className="face" data-test="face" onClick={() => dispatch(minesweeper.actions.resetGame())}>
        {gameStateToSmile[gameState]}
      </div>
      <Field />
    </div>
  );
};
