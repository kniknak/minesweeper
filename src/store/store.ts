import { configureStore, createSlice, PayloadAction } from "@reduxjs/toolkit";
import { CellState, GameState, MinesweeperState } from "./types";
import { Queue } from "./Queue";
import { getCellsAround } from "./getCellsAround";

const initialState: MinesweeperState = {
  width: 30,
  height: 16,
  minesCount: 99,
  gameState: GameState.Default,
  cells: [],
  mines: [],
  minesAround: [],
};

export const minesweeper = createSlice({
  name: "counter",
  initialState,
  reducers: {
    /**
     * sets field size & mines count
     */
    setSetting: (state, action: PayloadAction<{ field: "width" | "height" | "minesCount"; value: number }>) => {
      if (action.payload.field === "minesCount") {
        state[action.payload.field] = Math.max(action.payload.value, 1) || 1;
      } else {
        state[action.payload.field] = Math.max(action.payload.value, 2) || 2;
      }

      state.minesCount = Math.min(state.minesCount, state.width * state.height - 1);
    },
    /**
     * resets the game
     */
    resetGame: (state) => {
      state.gameState = GameState.Default;
      state.cells = [];
      state.mines = [];
      state.minesAround = [];
    },
    /**
     * starts the game from the given cell and places mines
     *
     * 1st cell is always safe to open
     */
    startGame: (state, action: PayloadAction<number>) => {
      console.time("start");

      const emptyArray = Array.from(Array(state.width * state.height));

      const mines = [...emptyArray].fill(false);

      const minesIndexes: number[] = [];

      while (minesIndexes.length < state.minesCount) {
        let mineIndex = action.payload;

        while (mineIndex === action.payload || minesIndexes.includes(mineIndex)) {
          mineIndex = Math.floor(Math.random() * state.width * state.height);
        }

        minesIndexes.push(mineIndex);
      }

      minesIndexes.forEach((id) => (mines[id] = true));

      state.mines = mines;
      state.cells = [...emptyArray].fill(CellState.Default);
      state.minesAround = [...emptyArray]
        .fill(0)
        .map((_, index) => getCellsAround(index, state.width, state.height).filter((id) => mines[id]).length);
      state.gameState = GameState.Started;

      console.timeEnd("start");
    },
    /**
     * adds or removes flag from the field
     */
    toggleFlag: (state, action: PayloadAction<number>) => {
      if (state.cells[action.payload] !== CellState.Revealed) {
        if (state.cells[action.payload] === CellState.Default) {
          if (state.cells.filter((cellState) => cellState === CellState.Flagged).length < state.minesCount) {
            state.cells[action.payload] = CellState.Flagged;
          }
        } else {
          state.cells[action.payload] = CellState.Default;
        }
      }
    },
    /**
     * checks the cell state
     *
     * recursively opens the nearest cells
     */
    revealCell: (state, action: PayloadAction<number>) => {
      console.time("revealCell");

      if (state.mines[action.payload]) {
        state.cells[action.payload] = CellState.Revealed;
        state.gameState = GameState.Failed;
      } else {
        const queue = new Queue();
        queue.push(action.payload);

        while (queue.size) {
          const id: number = queue.pop();

          state.cells[id] = CellState.Revealed;

          if (!state.minesAround[id]) {
            getCellsAround(id, state.width, state.height)
              .filter((id) => state.cells[id] === CellState.Default)
              .filter((id) => !queue.has(id))
              .forEach((id) => queue.push(id));
          }
        }
      }

      const hasSucceeded = state.cells.every((cellState, index) =>
        state.mines[index] ? cellState !== CellState.Revealed : cellState === CellState.Revealed,
      );

      if (hasSucceeded) {
        state.gameState = GameState.Succeeded;
      }

      console.timeEnd("revealCell");
    },
  },
});

export const store = configureStore({
  reducer: minesweeper.reducer,
});

export type AppDispatch = typeof store.dispatch;
