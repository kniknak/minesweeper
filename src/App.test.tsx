import React, { Dispatch } from "react";
import { shallow } from "enzyme";
import { useSelector } from "react-redux";
import { minesweeper } from "./store/store";
import { App } from "./App";
import { GameState } from "./store/types";

jest.mock("react-redux", () => ({
  useDispatch: () => (...args: any[]) => args,
  useSelector: jest.fn(),
}));

jest.mock("react", () => ({
  ...jest.requireActual<typeof import("react")>("react"),
  memo: (Component: React.FC<Record<string, any>>) => Component,
}));

jest.mock("./store/store", () => ({
  minesweeper: {
    actions: {
      resetGame: jest.fn(),
    },
  },
}));

const useSelectorMock = useSelector as jest.Mock<GameState>;

describe("App", () => {
  beforeEach(() => {
    useSelectorMock.mockReset();
  });

  it("renders", () => {
    useSelectorMock.mockReturnValue(GameState.Default);

    const subject = shallow(<App />);

    expect(subject).toMatchSnapshot();
  });

  it("resets game", () => {
    useSelectorMock.mockReturnValue(GameState.Default);

    const subject = shallow(<App />);

    subject.find('[data-test="face"]').prop("onClick")!({} as any);

    expect(minesweeper.actions.resetGame).toHaveBeenCalled();
  });
});
