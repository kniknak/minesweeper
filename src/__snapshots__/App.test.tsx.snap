// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<div
  className="app"
>
  <Settings />
  <div
    className="face"
    data-test="face"
    onClick={[Function]}
  >
    ðŸ™‚
  </div>
  <Field />
</div>
`;
